### ISSUE TITLE
nvidia.dali.plugin.tf makes my label incorrect

### ISSUE THREAD

**Issue by wangguangyuan** [NVIDIA/DALI#82]
When I use pipe.build() and pipe.run() to enter the result, the label output is correct, but when I use nvidia.dali.plugin.tf and then use sess.run() to output the result, label is all zero.
my test file format: (image_path, label)
/n01704323/n01704323_1569.JPEG 0
/n01704323/n01704323_5833.JPEG 0
/n01704323/n01704323_2956.JPEG 0
/n01704323/n01704323_4976.JPEG 0
/n01704323/n01704323_8911.JPEG 0
/n01704323/n01704323_9292.JPEG 0
/n01704323/n01704323_1379.JPEG 0
/n01704323/n01704323_8648.JPEG 0
/n01704323/n01704323_8497.JPEG 0
/n01704323/n01704323_2732.JPEG 0
/n02017213/n02017213_3476.JPEG 1
/n02017213/n02017213_6601.JPEG 1
/n02017213/n02017213_3582.JPEG 1
/n02017213/n02017213_765.JPEG 1
/n02017213/n02017213_3461.JPEG 1
/n02017213/n02017213_3894.JPEG 1
/n02017213/n02017213_513.JPEG 1
/n02017213/n02017213_6029.JPEG 1
/n02017213/n02017213_43.JPEG 1
/n02017213/n02017213_5432.JPEG 1
/n02017213/n02017213_7325.JPEG 1
/n02017213/n02017213_7622.JPEG 1
/n02017213/n02017213_7330.JPEG 1

my test code：
`
#First test procedure：
pipe = FileReadPipeline(batch_size=32, num_threads=1, device_id=0, random_shuffle=True,
file_list=save_file_path, initial_fill=4096)
pipe.build()
for _ in range(1):
pipe_out = pipe.run()
images, labels = pipe_out
labels_tensor = labels.asCPU().as_tensor()
images = images.asCPU()
print(np.array(labels_tensor))
# print(images.at(9))

output:
[[1]
[0]
[1]
[0]
[1]
[1]
[1]
[0]
[1]
[1]
[0]
[1]
[1]
[0]
[0]
[0]
[1]
[0]
[0]
[0]
[1]
[1]
[0]
[0]
[0]
[1]
[0]
[1]
[0]
[1]
[0]
[0]]
#--------------------------------------#
#Second test procedure:
device_id = 0
pipe = FileReadPipeline(batch_size=32, num_threads=1, device_id=0, random_shuffle=True,
file_list=save_file_path, initial_fill=4096)
seri_pipe = pipe.serialize()
daliop = dali_tf.DALIIterator()
with tf.device('/gpu:%i' % device_id):
image, label = daliop(serialized_pipeline=seri_pipe,
batch_size=32,
height=224,
width=224,
device_id=device_id)
gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=0.5)
config = tf.ConfigProto(gpu_options=gpu_options)
with tf.Session(config=config) as sess:
ims, labels = sess.run([image, label])
print(labels)
`
output:
[[ 0.00000000e+00]
[ 0.00000000e+00]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 0.00000000e+00]
[ 0.00000000e+00]
[ 1.40129846e-45]
[ 0.00000000e+00]
[ 0.00000000e+00]
[ 1.40129846e-45]
[ 0.00000000e+00]
[ 0.00000000e+00]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 0.00000000e+00]
[ 0.00000000e+00]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 1.40129846e-45]
[ 0.00000000e+00]
[ 0.00000000e+00]
[ 1.40129846e-45]
[ 0.00000000e+00]
[ 1.40129846e-45]
[ 0.00000000e+00]
[ 0.00000000e+00]
[ 0.00000000e+00]
[ 0.00000000e+00]]

`
class CommonPipeline(Pipeline):
def init(self, batch_size, num_threads, device_id, size=(224, 224), crop_size=(224, 224),
mean=IMAGE_MEAN,
std=IMAGE_STD,
channel_format=types.NHWC,
probability=0.5,
device='gpu',
decode_method='in_gpu'):
super(CommonPipeline, self).init(batch_size, num_threads, device_id)
assert device == 'gpu' and decode_method == 'in_gpu'
if decode_method == 'in_gpu':
self.decode = ops.nvJPEGDecoder(device="mixed", output_type=types.RGB)
else:
self.decode = ops.HostDecoder(output_type=types.RGB)

    self.rrc = ops.RandomResizedCrop(device=device, size=size)
    self.cmnp = ops.CropMirrorNormalize(device=device,
                                        output_dtype=types.FLOAT,
                                        output_layout=channel_format,
                                        image_type=types.RGB,
                                        crop=crop_size,
                                        mean=mean,
                                        std=std)
    self.coin = ops.CoinFlip(probability=probability)
    # self.cast = ops.Cast(device=device, dtype=types.FLOAT16)

def base_define_graph(self, inputs, labels):
    rng = self.coin()
    images = self.decode(inputs)
    images = self.rrc(images)
    images = self.cmnp(images, mirror=rng)
    # images = self.cast(images)
    return images, labels.gpu()
class FileReadPipeline(CommonPipeline):
def init(self,
batch_size,
num_threads,
device_id,
random_shuffle,
file_root='',
file_list='',
size=(224, 224),
crop_size=(224, 224),
mean=IMAGE_MEAN,
std=IMAGE_STD,
channel_format=types.NHWC,
probability=0.5,
device='gpu',
decode_method='in_gpu',
initial_fill=4096):
super(FileReadPipeline, self).init(batch_size=batch_size,
num_threads=num_threads,
device_id=device_id,
size=size,
crop_size=crop_size,
mean=mean,
std=std,
channel_format=channel_format,
probability=probability,
device=device,
decode_method=decode_method)
self.input = ops.FileReader(file_root=file_root, file_list=file_list,
random_shuffle=random_shuffle, initial_fill=initial_fill)

def define_graph(self):
    images, labels = self.input()
    return self.base_define_graph(images, labels)`

**Comment by cliffwoolley** [comment]
1.40129846e-45 is 0x1 reinterpreted as a float32 (rather than converted to float32).  @kh4l, any ideas here?

**Comment by JanuszL** [comment]
Tracked internally as DALI-180.

**Comment by Kh4L** [comment]
Hi  @wangguangyuan !
As @cliffwoolley mentioned, this is TensorFlow reinterpreting DALI outputs as float32.

This issue was addressed in https://github.com/NVIDIA/DALI/issues/28 , and fixed in https://github.com/NVIDIA/DALI/pull/33 .

However, `label_type` default type is `tf.int32`, so you shouldn't get floats.

Can you try adding `label_type = tf.int32` as an argument of `daliop`?

**Comment by wangguangyuan** [comment]
class ClassCommonPipeline(Pipeline):
    def __init__(self,
                 batch_size,
                 num_threads,
                 dali_device_id,
                 size=(224, 224),
                 crop_size=(224, 224),
                 seed=12,
                 mean=IMAGE_MEAN,
                 std=IMAGE_STD,
                 output_type=types.RGB,
                 channel_format=types.NHWC,
                 probability=0.5,
                 device='gpu',
                 decode_method='in_gpu'):

        super(ClassCommonPipeline, self).__init__(batch_size, num_threads, dali_device_id, seed=seed)

        assert (device == 'gpu' and decode_method == 'in_gpu') or (device == 'cpu' and decode_method == 'in_cpu')

        self.device = device
        self.decode_method = decode_method

        if decode_method == 'in_gpu':
            self.decode = ops.nvJPEGDecoder(device="mixed", output_type=output_type)
        else:
            self.decode = ops.HostDecoder(output_type=output_type)

        self.rrc = ops.RandomResizedCrop(device=device, size=size)

        self.cmnp = ops.CropMirrorNormalize(device=device,
                                            output_dtype=types.FLOAT,
                                            output_layout=channel_format,
                                            image_type=output_type,
                                            crop=crop_size,
                                            mean=mean,
                                            std=std)

        self.coin = ops.CoinFlip(probability=probability)
        # self.cast = ops.Cast(device=device, dtype=types.FLOAT16)

    def base_define_graph(self, inputs, labels):
        rng = self.coin()
        images = self.decode(inputs)
        images = self.rrc(images)
        images = self.cmnp(images, mirror=rng)
        # images = self.cast(images)
        if self.decode_method == 'in_gpu' and self.device == 'gpu':
            return images, labels.gpu()
        else:
            return images, labels


class ClassFileReadPipeline(ClassCommonPipeline):
    def __init__(self,
                 batch_size,
                 num_threads,
                 dali_device_id,
                 random_shuffle,
                 num_shards=1,
                 shard_id=0,
                 file_root='',
                 file_list='',
                 size=(224, 224),
                 crop_size=(224, 224),
                 seed=12,
                 mean=IMAGE_MEAN,
                 std=IMAGE_STD,
                 output_type=types.RGB,
                 channel_format=types.NHWC,
                 probability=0.5,
                 device='gpu',
                 decode_method='in_gpu',
                 initial_fill=4096):

        super(ClassFileReadPipeline, self).__init__(batch_size=batch_size,
                                                    num_threads=num_threads,
                                                    dali_device_id=dali_device_id,
                                                    size=size,
                                                    crop_size=crop_size,
                                                    seed=seed,
                                                    mean=mean,
                                                    std=std,
                                                    output_type=output_type,
                                                    channel_format=channel_format,
                                                    probability=probability,
                                                    device=device,
                                                    decode_method=decode_method)

        self.input = ops.FileReader(file_root=file_root, file_list=file_list,
                                    num_shards=num_shards, shard_id=shard_id,
                                    random_shuffle=random_shuffle, initial_fill=initial_fill)

    def define_graph(self):
        images, labels = self.input()
        return self.base_define_graph(images, labels)


class ClassTFRecordPipeline(ClassCommonPipeline):
    pass


def get_batch_images(pipe_name,
                     batch_size,
                     num_threads,
                     device_id,
                     random_shuffle,
                     dali_device_id,
                     shape,
                     num_shards=1,
                     shard_id=0,
                     file_root='',
                     file_list='',
                     size=(224, 224),
                     crop_size=(224, 224),
                     seed=12,
                     mean=IMAGE_MEAN,
                     std=IMAGE_STD,
                     output_type=types.RGB,
                     channel_format=types.NHWC,
                     probability=0.5,
                     device='gpu',
                     decode_method='in_gpu',
                     initial_fill=4096):

    pipes = pipe_name(batch_size=batch_size,
                      num_threads=num_threads,
                      dali_device_id=dali_device_id,
                      random_shuffle=random_shuffle,
                      num_shards=num_shards,
                      shard_id=shard_id,
                      file_root=file_root,
                      file_list=file_list,
                      size=size,
                      crop_size=crop_size,
                      seed=seed,
                      mean=mean,
                      std=std,
                      output_type=output_type,
                      channel_format=channel_format,
                      probability=probability,
                      device=device,
                      decode_method=decode_method,
                      initial_fill=initial_fill)

    print(pipes)
    serialized_pipes = pipes.serialize()

    daliop_t = dali_tf.DALIIterator()

    print('/gpu:%i' % device_id)
    with tf.device('/gpu:%i' % device_id):
        image, label = daliop_t(serialized_pipeline=serialized_pipes,
                                shape=shape,
                                image_type=tf.float32,
                                label_type=tf.int32,
                                num_threads=num_threads,
                                device_id=dali_device_id)

        return image, label


if __name__ == '__main__':
    root_path = '/share5/public/classification_data/imagenet1k/train'
    file_path = '/share5/public/classification_data/imagenet1k/meta/train.txt'
    save_path = '/share5/public/classification_data/imagenet1k/meta/absolute_train.txt'
    # transform_absolute_path(root_path, file_path, save_path)
    file_path2 = '/share5/public/guangyuan/workplace/horovod_project/imagenet'
    # get_image_and_label('', file_path='/share5/public/guangyuan/workplace/horovod_project/imagenet')
    # raise('dd')
    n_test = 30
    batch_size = 32
    n_threads = 4
    device_id = 0

    # pipe = ClassFileReadPipeline(batch_size=batch_size, num_threads=n_threads, dali_device_id=device_id, random_shuffle=True,
    #                              file_list=save_path, initial_fill=4096)
    #
    # pipe.build()
    # start_time = timer()
    # for _ in range(n_test):
    #     pipe_out = pipe.run()
    #     images, labels = pipe_out
    # t = timer() - start_time
    # print("Speed: {} imgs/s".format((n_test * batch_size) / t))


    # seri_pipe = pipe.serialize()
    # daliop = dali_tf.DALIIterator()
    #
    # with tf.device('/gpu:%i' % device_id):
    #     image, label = daliop(serialized_pipeline=seri_pipe,
    #                           batch_size=batch_size,
    #                           height=224,
    #                           width=224,
    #                           num_threads=4,
    #                           device_id=device_id)

    image, label = get_batch_images(ClassFileReadPipeline, batch_size, n_threads,
                                    device_id, False, dali_device_id=device_id, shape=[batch_size, 3, 224, 224],
                                    num_shards=8, shard_id=7,
                                    device='gpu', decode_method='in_gpu', channel_format=types.NCHW,
                                    file_list='./test.txt', initial_fill=4096)
    print(image, label)
    label = tf.squeeze(label)
    gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=0.7)
    config = tf.ConfigProto(gpu_options=gpu_options)
    with tf.Session(config=config) as sess:
        start_time = timer()
        for _ in range(n_test):
            ims, labs = sess.run([image, label])
            print(labs)
        # print(ims)
        # print(labs)
        t = timer() - start_time
        print("Speed: {} imgs/s".format((n_test * batch_size) / t))

**Comment by JanuszL** [comment]
Hi,
As I understand it works as you want it to?
I'm asking because I see labels with value 7 while you mentioned in your previous comments that 0 and 1 are available in your test code.

**Comment by wangguangyuan** [comment]
Yes, it works as I wish. I changed my labels to test the parameters of the FileReader num_shards and shard_id. 
**label_type** default type is tf.int32, so I shouldn't get floats.

**Comment by JanuszL** [comment]
Understood. Thank you for confirmation.
